version: "3.5"

services:
  api01: &api
    image: jjeanjacques10/rinhabackend-2024-q1:latest
    hostname: api01
    network_mode: host
    restart: on-failure
    ulimits:
      nofile:
        soft: 1000000
        hard: 1000000
    environment:
      - DB_HOSTNAME=localhost
      - DB_USER=admin
      - DB_PASSWORD=123
      - DB_INITIAL_POOL_SIZE=13
      - DB_MAX_POOL_SIZE=13
      - API_USE_DB_FUNC=true
      - LOG_LEVEL=INFO
      - PORT=8080
    depends_on:
      db:
        condition: service_healthy
    deploy:
      resources:
        limits:
          cpus: "0.60"
          memory: "150MB"

  api02:
    <<: *api
    hostname: api02
    environment:
      - PORT=8081

  nginx:
    image: nginx:latest
    network_mode: host
    restart: on-failure
    volumes:
      - ./nginx.conf:/etc/nginx/nginx.conf:ro
    depends_on:
      - api01
      - api02
    deploy:
      resources:
        limits:
          cpus: "0.10"
          memory: "20MB"

  db:
    image: postgres:latest
    hostname: db
    network_mode: host
    restart: on-failure
    environment:
      - POSTGRES_PASSWORD=123
      - POSTGRES_USER=admin
      - POSTGRES_DB=rinha
    healthcheck:
      test: [ 'CMD', 'pg_isready' ]
      interval: 10s
      timeout: 5s
      retries: 5
    volumes:
      - ./sql/init.sql:/docker-entrypoint-initdb.d/init.sql
      - ./postgresql.conf:/etc/postgresql/postgresql.conf
    command: postgres -c config_file=/etc/postgresql/postgresql.conf
    deploy:
      resources:
        limits:
          cpus: "0.20"
          memory: "230MB"